pipelines/controller/src/main/java/org/openmrs/analytics/PipelineManager.java

```diff
@@ -432,9 +493,60 @@ private void createHiveResourceTables(
     }
   }
 
+  /** This method captures the given exception into a file rooted at the dwhRoot location. */
+  void captureError(String dwhRoot, Exception e) {
+    try {
+      if (!Strings.isNullOrEmpty(dwhRoot)) {
+        String stackTrace = ExceptionUtils.getStackTrace(e);
+        DwhFiles.forRoot(dwhRoot)
+            .writeToFile(ERROR_FILE_NAME, stackTrace.getBytes(StandardCharsets.UTF_8));
+      }
+    } catch (IOException ex) {
+      logger.error("Error while capturing error to a file", ex);
+    }
+  }
+
+  /** Sets the details of the last pipeline run with the given dwhRoot as the snapshot location. */
+  void setLastRunDetails(String dwhRoot, Status status) {
+    DwhRunDetails dwhRunDetails = new DwhRunDetails();
+    try {
+      DwhFiles dwhFiles = DwhFiles.forRoot(dwhRoot);
+      String startTime = dwhFiles.readTimestampFile(DwhFiles.TIMESTAMP_FILE_START).toString();
+      dwhRunDetails.setStartTime(startTime);
+      if (status == Status.SUCCESS) {
+        String endTime = dwhFiles.readTimestampFile(DwhFiles.TIMESTAMP_FILE_END).toString();
+        dwhRunDetails.setEndTime(endTime);
+      } else {
+        dwhRoot = dwhRoot.endsWith("/") ? dwhRoot : dwhRoot + "/";
+        dwhRunDetails.setLogFilePath(dwhRoot + ERROR_FILE_NAME);
+      }
+      dwhRunDetails.setStatus(status);
+      dwhRunDetails.setParquetPath(dwhRoot);
+      this.lastRunDetails = dwhRunDetails;
+    } catch (IOException e) {
+      logger.error("Error in reading timestamp files", e);
+      throw new RuntimeException(e);
+    }
+  }
+
   public enum LastRunStatus {
     NOT_RUN,
     SUCCESS,
     FAILURE
   }
+
+  @Data
+  public class DwhRunDetails {
+
+    private String startTime;
+    private String endTime;
+    private Status status;
```

Comment:
nit: While in general I prefer to use `enum` over string constants, but for this case it seems to me that the main use for this value is to be read on the JS side. If that's true then maybe using a string constant (or maybe just a boolean) is better as we don't need to export package names in the JS code.